crates/runix/src/flake_ref2/mod.rs

```diff
@@ -0,0 +1,290 @@
+use std::borrow::Cow;
+use std::fmt::Display;
+use std::str::FromStr;
+
+use chrono::{NaiveDateTime, TimeZone, Utc};
+use derive_more::{Display, From};
+use serde::{Deserialize, Deserializer, Serialize};
+use thiserror::Error;
+
+use self::file::FileRef;
+use self::git::GitRef;
+use self::git_service::{service, GitServiceRef};
+use self::indirect::IndirectRef;
+use self::path::PathRef;
+
+pub mod file;
+pub mod git;
+pub mod git_service;
+pub mod indirect;
+pub mod lock;
+pub mod path;
+pub mod protocol;
+
+pub trait FlakeRefSource: FromStr + Display {
+    fn scheme() -> Cow<'static, str>;
+    fn parses(maybe_ref: &str) -> bool {
+        maybe_ref.starts_with(&format!("{}:", Self::scheme()))
+    }
+}
+
+#[derive(Serialize, Deserialize, Display, From, Debug, Clone, PartialEq, Eq)]
+#[serde(untagged)]
+pub enum FlakeRef {
+    FileFile(FileRef<protocol::File, file::File>),
+    FileHTTP(FileRef<protocol::HTTP, file::File>),
+    FileHTTPS(FileRef<protocol::HTTPS, file::File>),
+    TarballFile(FileRef<protocol::File, file::Tarball>),
+    TarballHTTP(FileRef<protocol::HTTP, file::Tarball>),
+    TarballHTTPS(FileRef<protocol::HTTPS, file::Tarball>),
+    Github(GitServiceRef<service::Github>),
+    Gitlab(GitServiceRef<service::Gitlab>),
+    Path(PathRef),
+    GitPath(GitRef<protocol::File>),
+    GitSsh(GitRef<protocol::SSH>),
+    GitHttps(GitRef<protocol::HTTPS>),
+    GitHttp(GitRef<protocol::HTTP>),
+    Indirect(IndirectRef),
+    // /// https://cs.github.com/NixOS/nix/blob/f225f4307662fe9a57543d0c86c28aa9fddaf0d2/src/libfetchers/tarball.cc#L206
+    // Tarball(TarballRef),
+}
+
+impl FromStr for FlakeRef {
+    type Err = ParseFlakeRefError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let flake_ref = match s {
+            _ if FileRef::<protocol::File>::parses(s) => {
```

Comment:
This feels a little more bug prone than matching based on schema, because all the `parses()` functions aren't necessarily mutually exclusive, and if one of them has a bug it will potentially break every match arm after it. Maybe there's enough complexity here to make that necessary, but I do wonder if the previous approach was a bit more readable and less bug prone

Reply:
thing is there is more complexity than just matching the prefix.

- see #16 for file and tarball
- we need to support `./path` and `/path` as well as `nixpkgs` -> `flake:nixpkgs`


sure if the first arm says "yup i take that" and then cant parse it, that will break the correct parser

i did have the issue with everything that failed parsing for any reason got parsed by `IndirectRef` before [ca47c9d](https://github.com/flox/runix/pull/15/commits/ca47c9d59354649682a27887eed2f275ef52d582)

So yeah the parses methods have to be strict enough to not intersect and also more flexible than just match statements